src/gc-arena/src/context.rs

```diff
@@ -344,17 +351,27 @@ impl Context {
         ptr
     }
 
+    #[inline]
     fn write_barrier(&self, gc_box: GcBox) {
         // During the propagating phase, if we are mutating a black object, we may add a white
         // object to it and invalidate the invariant that black objects may not point to white
         // objects. Turn black obejcts to gray to prevent this.
         let header = gc_box.header();
         if self.phase.get() == Phase::Propagate && header.color() == GcColor::Black {
             header.set_color(GcColor::Gray);
-            self.gray_again.borrow_mut().push(gc_box);
+
+            // Outline the actual enqueueing code (which is somewhat expensive and won't be
+            // executed often) to promote the inlining of the write barrier.
+            #[cold]
+            fn enqueue(this: &Context, gc_box: GcBox) {
+                this.gray_again.borrow_mut().push(gc_box);
+            }
```

Comment:
It's probably very minor, but should `header.set_color(GcColor::Gray)` also be in `enqueue`?